components/nimbus/src/lib.rs

```diff
@@ -36,22 +37,24 @@ use once_cell::sync::OnceCell;
 use persistence::{Database, StoreId, Writer};
 use serde_derive::*;
 use serde_json::{Map, Value};
-use std::collections::HashSet;
 use std::path::PathBuf;
 use std::sync::Mutex;
+use std::{collections::HashSet, str::FromStr};
 use updating::{read_and_remove_pending_experiments, write_pending_experiments};
 use uuid::Uuid;
 
 const DEFAULT_TOTAL_BUCKETS: u32 = 10000;
 const DB_KEY_NIMBUS_ID: &str = "nimbus-id";
-
+pub const DB_KEY_INSTALLATION_DATE: &str = "installation-date";
 // The main `NimbusClient` struct must not expose any methods that make an `&mut self`,
 // in order to be compatible with the uniffi's requirements on objects. This is a helper
 // struct to contain the bits that do actually need to be mutable, so they can be
 // protected by a Mutex.
 #[derive(Default)]
 struct InternalMutableState {
     available_randomization_units: AvailableRandomizationUnits,
+    installation_date: Option<DateTime<Utc>>,
+    app_context: AppContext,
```

Comment:
This is probably my biggest annoyance with this change. The app context becomes mutable..

The purpose of this is to pass down the `days_since_install` all the way down to the targeting evaluation, and adding it to the context was the *fastest* way to do that. Alternatively, we can expose the concept of "first install" to the `enrollment` and `evaluator` modules, but that seemed like it would be a ton messier, and those modules have nothing to do with installation dates.

Reply:
Maybe a silly question, but: Could we just add the "installation_date" inside the `AppContext`? That way it provides a place for the app to provide this info, and if it's missing on init, then we can try to derive it ourselves. I guess the AppContext would still need to be mutable or replaced if we had to derive it, but then it's already there for when we pass it into the jexl-rs `evaluate_in_context` in `fn targeting`?